fero/analysis.py

```diff
@@ -148,3 +328,314 @@ def make_prediction(
+
+        if is_cost:
+            for factor in goal["cost_function"]:
+                cost_factors.append(factor["name"])
+        else:
+            if goal["factor"]["name"] in self.factor_names:
+                goal_factor.append(goal["goal"])
+            else:
+                target_factor.append(goal["goal"])
+
+        for constraint in constraints:
+            if constraint["name"] in self.factor_names:
+                constraint_factors.append(constraint["name"])
+            else:
+                constraint_targets.append(constraint["name"])
+
+        if len(target_factor + constraint_targets) < 1:
+            raise FeroError("No Targets specified")
+
+        if len(constraint_factors + cost_factors + goal_factor) > 3:
+            raise FeroError(
+                "A maximum of three factors can be specified in an optimization"
+            )
+
+    def _verify_fixed_factors(self, fixed_factors: dict):
+        """Check that the provided fixed factors are in the analysis"""
+        all_columns = self.target_names + self.factor_names
+        for key in fixed_factors.keys():
+            if key not in all_columns:
+                raise FeroError(f'"{key}" is not a valid factor')
+
+    def _get_basis_values(self):
+        """Gets median fixed values from presentation data"""
+        reg_data = next(
+            d
+            for d in self._presentation_data["data"]
+            if d["id"] == "regression_simulator"
+        )
+        # yuck massive ugly json
+        base_values = {f["factor"]: f["median"] for f in reg_data["content"]["factors"]}
+
+        target_values = {
+            t["name"]: t["default"]["mid"][0] for t in reg_data["content"]["targets"]
+        }
+
+        base_values.update(target_values)
+        return base_values
+
+    def _build_optimize_request(
+        self, name, goal, constraints, fixed_factors, include_confidence_intervals
+    ) -> dict:
+        """Formats the content for an optimization request"""
+
+        is_cost = goal.get("type", None) == "COST"
+        ci_value = "include" if include_confidence_intervals else "exclude"
+        opt_request = {
+            "name": name,
+            "description": "",
+            "prediction_type": "O",
+        }
+
+        input_data = {
+            "kind": "OptimizationRequest",
+            "version": 1,
+            "objectives": [
+                {
+                    "factor": FERO_COST_FUNCTION if is_cost else goal["factor"]["name"],
+                    "goal": goal["goal"],
+                }
+            ],
+            "basisSpecifiedColumns": [],  # These appear to just be, uhm, here
+            "linearFunctionDefinitions": {},
+        }
+        basis_values = self._get_basis_values()
+        basis_values.update(fixed_factors)
+        input_data["basisValues"] = basis_values
+
+        factor_bounds = [
+            {
+                "factor": c["name"],
+                "lowerBound": c["min"],
```

Comment:
Should just make sure that this doesn't blow up on a categorical factor. They don't have bounds. `None` should be okay, but is probably worth checking.